pkg/controller/statefulset/stateful_set.go

```diff
@@ -309,11 +339,23 @@ func (ssc *StatefulSetController) deletePod(logger klog.Logger, obj interface{})
 // NOTE: Returned Pods are pointers to objects from the cache.
 // If you need to modify one, you need to copy it first.
 func (ssc *StatefulSetController) getPodsForStatefulSet(ctx context.Context, set *apps.StatefulSet, selector labels.Selector) ([]*v1.Pod, error) {
-	// List all pods to include the pods that don't match the selector anymore but
-	// has a ControllerRef pointing to this StatefulSet.
-	pods, err := ssc.podLister.Pods(set.Namespace).List(labels.Everything())
-	if err != nil {
-		return nil, err
+	// Iterate over two keys:
+	//  The UID of the StatefulSet, which identifies Pods that are controlled by the StatefulSet.
+	//  The OrphanPodIndexKey, which helps identify orphaned Pods that are not currently managed by any controller,
+	//   but may be adopted later on if they have matching labels with the StatefulSet.
+	podsForSts := []*v1.Pod{}
+	for _, key := range []string{string(set.UID), OrphanPodIndexKey} {
+		if podObjs, err := ssc.podIndexer.ByIndex(StatefulSetIndex, key); err == nil {
+			for _, obj := range podObjs {
+				if pod, ok := obj.(*v1.Pod); ok {
+					podsForSts = append(podsForSts, pod)
+				} else {
```

Comment:
To reduce the number of indentation and improve readability, I'd suggest handling the error or !ok case first:
```
pod, ok := obj.(*v1.Pod)
if !ok {
    // Do something
}
podsForSts = append(podsForSts, pod)
```